### PR TITLE
[test] Modernize PETSc test

### PR DESCRIPTION
**Pull request by teojgo**

### CODE REVIEW

**Review by kraushm — APPROVED** [review]
lgtm.

File: cscs-checks/libraries/petsc/petsc_helloworld.py [hunk 1]
```diff
@@ -7,30 +7,39 @@
...
-        self.num_tasks_per_node = 8
-        self.build_system = 'SingleSource'
+                      f'({self.linkage} linking)')
+
+    @run_after('setup')
+    def set_variables(self):
+        self.variables = {'CRAYPE_LINK_TYPE': self.linkage}
+
+    @run_before('compile')
+    def prg_intel_workaround(self):
```
Comment (vkarak) [inline cscs-checks/libraries/petsc/petsc_helloworld.py:34]:
```suggestion
    def intel_workaround(self):
```

File: cscs-checks/libraries/petsc/petsc_helloworld.py [hunk 2]
```diff
@@ -7,30 +7,39 @@
...
-        self.valid_prog_environs = ['PrgEnv-cray', 'PrgEnv-gnu',
-                                    'PrgEnv-intel']
-        self.sourcepath = 'poisson2d.c'
-        self.modules = ['cray-petsc']
-        self.num_tasks = 16
-        self.num_tasks_per_node = 8
-        self.build_system = 'SingleSource'
+                      f'({self.linkage} linking)')
+
+    @run_after('setup')
```
Comment (vkarak) [inline cscs-checks/libraries/petsc/petsc_helloworld.py:29]: I would put that in a pre-compile hook.

**Review by vkarak — CHANGES_REQUESTED** [review]

**Review by vkarak — APPROVED** [review]

**Comment by teojgo** [comment]
@jenkins-cscs retry daint

**Pull request merged**